solidity/test/submitERC721LogicCall.ts

```diff
@@ -0,0 +1,448 @@
+import chai from "chai";
+import { ethers } from "hardhat";
+import { solidity } from "ethereum-waffle";
+import { TestERC721A } from "../typechain/TestERC721A";
+import { GravityERC721 } from "../typechain/GravityERC721";
+import { deployContracts } from "../test-utils/deployERC721";
+import {
+  getSignerAddresses,
+  signHash,
+  examplePowers,
+  ZeroAddress
+} from "../test-utils/pure";
+import { ERC721, TestERC20A } from "../typechain";
+
+chai.use(solidity);
+const { expect } = chai;
+
+
+async function runTest(opts: {
+  // Issues with the tx batch
+  invalidationNonceNotHigher?: boolean;
+  malformedTxBatch?: boolean;
+
+  // Issues with the current valset and signatures
+  nonMatchingCurrentValset?: boolean;
+  badValidatorSig?: boolean;
+  zeroedValidatorSig?: boolean;
+  notEnoughPower?: boolean;
+  barelyEnoughPower?: boolean;
+  malformedCurrentValset?: boolean;
+  timedOut?: boolean;
+  wrongOwner?: boolean;
+}) {
+
+  // Prep and deploy contract
+  // ========================
+  const signers = await ethers.getSigners();
+  const gravityId = ethers.utils.formatBytes32String("foo");
+  // This is the power distribution on the Cosmos hub as of 7/14/2020
+  let powers = examplePowers();
+  let validators = signers.slice(0, powers.length);
+  const {
+    gravity,
+    gravityERC721,
+    testERC721,
+    testERC20,
+    checkpoint: deployCheckpoint
+  } = await deployContracts(gravityId, validators, powers);
+
+
+  const TestGravityERC721Contract = await ethers.getContractFactory("GravityERC721");
+  const ERC721LogicContract = (await TestGravityERC721Contract.deploy(gravity.address)) as GravityERC721;
+  // We set its owner to the batch contract. 
+  await gravityERC721.transferOwnership(gravity.address);
+
+  // Transfer out to Cosmos, locking coins
+  // =====================================
+  await testERC20.functions.approve(gravity.address, 1000);
+  await gravity.functions.sendToCosmos(
+    testERC20.address,
+    ethers.utils.formatBytes32String("myCosmosAddress"),
+    1000
+  );
+
+  const numTxs = 10;
+  const txPayloads = new Array(numTxs);
+  const txAmounts = new Array(numTxs);
+  const tokenIds = new Array(numTxs);
+  const destinations = new Array(numTxs);
+
+  for (let i = 0; i < numTxs; i++) {
+    await testERC721.functions.approve(gravityERC721.address, 1+i);
+    await gravityERC721.functions["sendERC721ToCosmos(address,string,uint256)"](
+      testERC721.address,
+      ethers.utils.formatBytes32String("myCosmosAddress"),
+      1+i);
+    tokenIds[i] = 1+i;
+    destinations[i] = signers[i + 5].address;
+    txAmounts[i] = 0;
+  }
+
+  let invalidationNonce = 1
+  if (opts.invalidationNonceNotHigher) {
+    invalidationNonce = 0
+  }
+
+  let timeOut = 4766922941000
+  if (opts.timedOut) {
+    timeOut = 0
+  }
+
+
+  // // // Call method
+  // // // ===========
+  // // // We have to give the logicBatch contract 5 coins for each tx, since it will transfer that
+  // // // much to the logic contract.
```

Comment:
I think this is incorrect, 5 coins would be for the relayer reward right?

Reply:
comment updated